### PR TITLE
fix: install latest solc if no pragma-specs found

### PR DESCRIPTION
**Pull request by helloibis**
### What I did

Add logic to install the latest available Solc version if no contracts being compiled include a version

This is specific to a situation where the user is setting up a new project and has not installed Solc themselves

fixes: #51 

### How to verify it

Remove Solc installations (I just `rm -rf ~/.solcx/` since I didn't need any right now) and follow steps in #51. With this change it should no longer fail but instead leave you with the latest version of Solidity installed (solc-v0.8.15 in my case)

A test `test_compile_missing_version()` was also added to verify that a system without solidity installed will have the latest version downloaded if the contract being compiled does not specify a version

### Checklist

- [x] Passes all linting checks (pre-commit and CI jobs)
- [x] New test cases have been added and are passing
- [x] Documentation has been updated
- [x] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

File: ape_solidity/compiler.py [hunk 1]
```diff
@@ -423,6 +423,11 @@ def _get_pragma_spec(path: Path) -> Optional[NpmSpec]:
         # Build map of pragma-specs.
         source_by_pragma_spec = {p: _get_pragma_spec(p) for p in source_paths_to_compile}
 
+        # If not Solidity version has been auto-installed while fetching the
+        # contract versions, install the latest
+        if self.installed_versions:
```
Comment (unparalleled-js) [inline ape_solidity/compiler.py:428]: should this be `if not self.installed_versions:` ?

**Review by unparalleled-js — COMMENTED** [review]
One way I would look into testing this would be to create a context manager pytest fixture that temporary moves all the .solc installs and puts them back at the end of the test. So `solcx` will think its not installed and then install it

File: ape_solidity/compiler.py (hunk 1 above)
Comment (helloibis) [reply ape_solidity/compiler.py:428]: Dang it, I had that but forgot to save before pushing 😂

Comment (helloibis) [reply ape_solidity/compiler.py]: Just force pushed to correct

File: ape_solidity/compiler.py [hunk 2]
```diff
@@ -423,6 +423,11 @@ def _get_pragma_spec(path: Path) -> Optional[NpmSpec]:
         # Build map of pragma-specs.
         source_by_pragma_spec = {p: _get_pragma_spec(p) for p in source_paths_to_compile}
 
+        # If not Solidity version has been auto-installed while fetching the
+        # contract versions, install the latest
+        if not self.installed_versions:
```
[the remark this answers is not in the document]
Comment (helloibis) [reply ape_solidity/compiler.py]: Agree, although slightly tweaked logic: 

```python
if not self.installed_versions and not any(source_by_pragma_spec.values()):
    ...
```

If a contract does not define its version it ends up as `{<path>: None}` in the mapping. I don't believe any `Version` should evaluate to a falsey value so this should do the trick 

Updated in 241a021dc

File: ape_solidity/compiler.py [hunk 3]
```diff
@@ -423,6 +423,11 @@ def _get_pragma_spec(path: Path) -> Optional[NpmSpec]:
         # Build map of pragma-specs.
         source_by_pragma_spec = {p: _get_pragma_spec(p) for p in source_paths_to_compile}
 
+        # If not Solidity version has been auto-installed while fetching the
+        # contract versions, we must install the latest compiler
```
Comment (fubuloubu) [inline ape_solidity/compiler.py:427]:
```suggestion
        # If no Solidity version has been installed previously while fetching the
        # contract version pragma, we must install a compiler, so choose the latest
```

**Review by fubuloubu — APPROVED** [review]

**Comment by helloibis** [comment]
> One way I would look into testing this would be to create a context manager pytest fixture that temporary moves all the .solc installs and puts them back at the end of the test. So `solcx` will think its not installed and then install it

Yeah good call! I'll check this out tomorrow and try utilizing it for a new test against a missing compiler. Looks like there's already the `MissingPragma.sol` contract that can be used too for that kind of test

**Review by fubuloubu — APPROVED** [review]

**Comment by helloibis** [comment]
Just rebased off of Jules' fix for the plugin not installing solidity https://github.com/ApeWorX/ape-solidity/pull/55

Going to check out the test suite and see about getting an isolated filesystem so the tests manage their own solc versions and don't leave anything behind after running

**Review by fubuloubu — APPROVED** [review]

**Comment by helloibis** [comment]
Just updated to add `test_compile_missing_version()` which tests against a "clean" solidity installation (i.e. no compilers at the configured solcx path)

I was thinking this was something that could be done for each test to keep them isolated from each other and clean up what gets installed by the tests, but for every test that ended up taking like 12 minutes of run time so I left it only on the relevant test for now

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**